File: sdks/node/examples/direct.ts
// llmux DIRECT mode on Node — the gateway inside this process, over the C ABI.
//
//   npm run build && node examples/direct.ts
//
// No API key and no network: the example spawns examples/fake-upstream.mjs,
// which prints an llmux config routing the model "demo" at itself.
//
// Direct mode on Node is synchronous — every call here blocks the event loop,
// which this example measures rather than asserts. Read ../README.md before
// choosing it for a server; examples/sidecar.ts is the one to copy for that.

import { spawn } from "node:child_process";
import * as readline from "node:readline";
import { fileURLToPath } from "node:url";
import * as path from "node:path";

import { Gateway, abiVersion, resolveLibrary } from "../direct.js";

const here = path.dirname(fileURLToPath(import.meta.url));

/** Start the fake upstream and return its llmux config plus a kill function. */
async function startUpstream(
  text = "the quick brown fox jumps over the lazy dog",
  delayMs: number | string = process.env.FAKE_DELAY_MS ?? "40"
): Promise<{ config: string; stop: () => void }> {
  const child = spawn(process.execPath, [path.join(here, "fake-upstream.mjs")], {
    stdio: ["ignore", "pipe", "inherit"],
    env: { ...process.env, FAKE_TEXT: text, FAKE_DELAY_MS: String(delayMs) },
  });
  const rl = readline.createInterface({ input: child.stdout });
  const first = await rl[Symbol.asyncIterator]().next();
  if (first.done) throw new Error("fake upstream exited before printing its config");
  return { config: first.value.slice("CONFIG ".length), stop: () => void child.kill() };
}

/**
 * The base URL fake-upstream.mjs is actually listening on, derived from the
 * CONFIG line it printed rather than tracked separately — that config is the
 * one source of truth every example already reads, and a second copy of the
 * port would just be one more place for the two to drift.
 */
function harnessBaseURL(configJson: string): string {
  const cfg = JSON.parse(configJson) as { providers: { base_url: string }[] };
  const baseUrl = cfg.providers[0]?.base_url;
  if (!baseUrl) throw new Error("fake upstream config had no providers[0].base_url");
  return baseUrl.endsWith("/v1") ? baseUrl.slice(0, -"/v1".length) : baseUrl;
}

/** GET /generated on the harness: chunks actually written to a socket. */
async function generatedCount(harnessBase: string): Promise<{ generated: number; streams: number; disconnects: number }> {
  const res = await fetch(`${harnessBase}/generated`);
  return (await res.json()) as { generated: number; streams: number; disconnects: number };
}

async function main(): Promise<void> {
  console.log(`node       ${process.version} on ${process.platform}/${process.arch}`);
  console.log(`library    ${resolveLibrary()}`);
  console.log(`abi        ${abiVersion()}\n`);

  const upstream = await startUpstream();
  try {
    // `using` closes the handle on every exit path out of this block, including
    // a throw. Without it, an exception between open and close leaks a gateway.
    using gw = Gateway.open({ config: upstream.config, expectVersion: abiVersion() });
    console.log(`handle     ${String(gw.handle)}\n`);

    // ---- models: answered from memory, no upstream involved ----------------
    const models = gw.call("models") as { data: { id: string }[] };
    console.log("models      ", models.data.map((m) => m.id).join(", "));

    // ---- unary chat, and what it costs the event loop ----------------------
    // The upstream fixture sleeps 40 ms per word, so this takes ~400 ms. A 1 ms
    // interval runs across it: the tick count is how much of the event loop
    // survived. It is zero, and the honesty of this module rests on saying so.
    let ticks = 0;
    const timer = setInterval(() => ticks++, 1);
    const startedAt = Date.now();
    const answer = gw.call("chat", {
      model: "demo",
      messages: [{ role: "user", content: "hello" }],
    }) as { choices: { message: { content: string } }[] };
    const elapsed = Date.now() - startedAt;
    clearInterval(timer);
    console.log("chat        ", JSON.stringify(answer.choices[0]?.message.content));
    console.log(`             blocked the event loop for ${String(elapsed)} ms; timer fired ${String(ticks)}x\n`);

    // ---- streaming: one callback per chunk, on this thread -----------------
    process.stdout.write("stream      ");
    const delivered = gw.stream({ model: "demo", messages: [{ role: "user", content: "hello" }] }, (chunk) => {
      process.stdout.write(chunk.choices?.[0]?.delta?.content ?? "");
    });
    console.log(`\n             ${String(delivered)} chunks\n`);

    // ---- stopping early ----------------------------------------------------
    // Returning false returns non-zero from the C callback. llmux stops at the
    // next chunk boundary, llmux_stream still returns 0, and no error is raised:
    // you decided to stop, so llmux does not hand your own decision back to you.
    // The callback form cannot overrun. There is no queue between the library
    // and you, so the number of chunks the C callback delivered IS the number
    // your callback saw — Node's one advantage over the buffered async
    // iterators the Deno and Bun SDKs expose. Both numbers are printed so the
    // claim is checkable rather than asserted.
    let seen = 0;
    const delivered2 = gw.stream({ model: "demo", messages: [{ role: "user", content: "hello" }] }, () => ++seen < 3);
    console.log(`break       consumed ${String(seen)} chunks; the C callback fired ${String(delivered2)}x (10 = the whole answer)`);
    console.log("            no error raised — stopping is your decision, not a failure\n");

    // ---- cancellation: llmux_cancel via AbortSignal -------------------------
    // `break`, above, is a DIFFERENT feature: the consumer's own callback says
    // stop and llmux_stream honours it at the next boundary — an ordinary,
    // successful early exit. This section is about the other case: someone
    // ELSE deciding the call should die — a timeout, a request that was itself
    // cancelled, a supervisor giving up — reaching in from outside and making
    // llmux actually stop generating upstream, not just stop handing you
    // chunks. Before llmux 0.1.5 there was no way to do that without
    // llmux_close, which tears down the whole gateway and every other stream
    // on it.
    //
    // A dedicated upstream: real per-chunk delay and the /generated counter
    // fake-upstream.mjs exposes, both needed to prove the claim rather than
    // assume it. Nothing about the consumer's own chunk count shows the
    // provider stopped — only the upstream knows that, so the upstream is
    // asked.
    const cancelUpstream = await startUpstream("one two three four five six seven eight nine ten", 100);
    try {
      const cancelBase = harnessBaseURL(cancelUpstream.config);

      // Baseline: run to completion, uncancelled, so "N of a possible M" below
      // has a measured M instead of an assumed one. 10 words plus the trailing
      // empty-delta "stop" chunk llmux always sends is 11 chunks; this fixture
      // emits no usage frame (see fake-upstream.mjs), which is why this number
      // is 11 and not the 12 the Python harness reports for the same text.
      using full = Gateway.open({ config: cancelUpstream.config });
      let seenFull = 0;
      const deliveredFull = full.stream({ model: "demo", messages: [{ role: "user", content: "hello" }] }, () => {
        seenFull++;
      });
      const generatedFull = await generatedCount(cancelBase);
      console.log(
        `cancel      baseline:  consumed ${String(seenFull)} chunks (callback fired ${String(deliveredFull)}x), upstream generated ${String(generatedFull.generated)}`
      );

      // Cancelled: an AbortController whose abort() call happens INSIDE the
      // chunk callback — the only place a single-threaded host can reach it
      // while llmux_stream is blocking this thread. The 'abort' event on an
      // AbortSignal fires SYNCHRONOUSLY, so this reaches Gateway.cancel, and
      // therefore llmux_cancel, on the same call stack: before onChunk
      // returns, before llmux_stream's blocking call unwinds. Measured safe —
      // it does not deadlock, unlike calling close() from inside a callback.
      using partial = Gateway.open({ config: cancelUpstream.config });
      const ac = new AbortController();
      let seenPartial = 0;
      try {
        partial.stream(
          { model: "demo", messages: [{ role: "user", content: "hello" }] },
          () => {
            seenPartial++;
            if (seenPartial === 3) ac.abort();
          },
          { signal: ac.signal }
        );
        console.log("cancel      UNEXPECTED: stream() returned normally after cancel");
      } catch (e) {
        // A cancelled stream is a FAILURE return from llmux_stream — rc=-1,
        // *err="context canceled" — unlike returning non-zero from onChunk,
        // which is 0/no-error because that was the caller's own decision
        // handed back as an ordinary stop. Cancellation is llmux noticing the
        // caller lost interest, and it is reported as the error it is.
        //
        // /generated is cumulative for the life of this upstream process —
        // it already counted the 11 chunks the baseline run above produced —
        // so what THIS run generated is the delta since that measurement, not
        // the raw number.
        const generatedAfter = await generatedCount(cancelBase);
        const generatedThisRun = generatedAfter.generated - generatedFull.generated;
        console.log(
          `            cancelled: consumed ${String(seenPartial)} chunks, upstream generated ${String(generatedThisRun)} of ${String(generatedFull.generated)}`
        );
        console.log(`            error: ${e instanceof Error ? e.message : "non-Error thrown"}\n`);
      }

      // ---- the setTimeout trap ----------------------------------------------
      // llmux_stream blocks the event loop for the whole call, so a timer
      // armed BEFORE the call cannot run until the call returns — Node cannot
      // preempt its own main thread to service it. This is not a corner case
      // to work around; it is the direct-mode trade-off this package exists to
      // be honest about. Proof, not assertion: arm a 50 ms abort against a
      // stream that takes over a second and record when the listener actually
      // ran relative to when the stream started.
      using timed = Gateway.open({ config: cancelUpstream.config });
      const tac = new AbortController();
      const t0 = Date.now();
      let firedAt = -1;
      tac.signal.addEventListener("abort", () => {
        firedAt = Date.now() - t0;
      });
      setTimeout(() => {
        tac.abort();
      }, 50);
      let seenTimed = 0;
      const deliveredTimed = timed.stream(
        { model: "demo", messages: [{ role: "user", content: "hello" }] },
        () => {
          seenTimed++;
        },
        { signal: tac.signal }
      );
      const elapsed = Date.now() - t0;
      // Give the timers phase a turn. A bare setImmediate() is NOT enough here
      // — measured: it runs its check-phase callback before libuv gets back
      // to the overdue timers phase when a koffi callback registration is
      // still in the picture, so `firedAt` would read -1 even though the
      // timer is about to fire a tick later. A real (if tiny) setTimeout
      // forces a pass through the timers phase first. Either way, nothing
      // about the numbers above changes: every chunk was already delivered
      // before this line runs — this wait only lets the already-overdue
      // listener get scheduled so this example can report when it ran.
      await new Promise((r) => setTimeout(r, 10));
      console.log(
        `timer       50 ms abort armed against a ${String(elapsed)} ms stream; delivered ${String(seenTimed)} chunks (callback fired ${String(deliveredTimed)}x) unharmed`
      );
      console.log(
        `            listener fired ${firedAt < 0 ? "not yet" : `at ${String(firedAt)} ms`} — after the stream ended, not during it`
      );
      console.log("            a timer cannot preempt the blocked event loop; only an abort from inside onChunk can\n");
    } finally {
      cancelUpstream.stop();
    }

    // ---- the error path ----------------------------------------------------
    try {
      gw.call("chat", { model: "no-such-model", messages: [{ role: "user", content: "x" }] });
      console.log("error       UNEXPECTED: unknown model was accepted");
    } catch (e) {
      // Plain UTF-8 text, not JSON. The string was freed with llmux_free before
      // this line ran — see takeError() in ../direct.ts.
      console.log(`error       ${e instanceof Error ? e.message : "non-Error thrown"}`);
    }

    // ---- streaming refuses to be a unary call ------------------------------
    try {
      gw.call("chat", { model: "demo", messages: [{ role: "user", content: "x" }], stream: true });
      console.log("stream:true UNEXPECTED: accepted by llmux_call");
    } catch (e) {
      console.log(`stream:true ${e instanceof Error ? e.message : "non-Error thrown"}`);
    }

    // ---- a closed handle ---------------------------------------------------
    const doomed = Gateway.open({ config: upstream.config });
    doomed.close();
    doomed.close(); // idempotent, as llmux_close is
    try {
      doomed.call("models");
    } catch (e) {
      console.log(`closed      ${e instanceof Error ? e.message : "non-Error thrown"}`);
    }
  } finally {
    upstream.stop();
  }
}

await main();

File: sdks/rust/examples/direct.rs
//! llmux in this process, over the C ABI.
//!
//! ```text
//! cargo run --example direct                  # needs provider keys in the env
//! ../../sdks/rust/examples/run.sh direct      # offline, against ffi/fakeupstream
//! ```
//!
//! Environment:
//!   LLMUX_LIBRARY      path to libllmux.dylib / .so (else it is searched for)
//!   LLMUX_CONFIG_JSON  an llmux config document; unset means defaults + env
//!   LLMUX_MODEL        model to route (default "demo")

use std::process::ExitCode;
use std::time::Instant;

use llmux::direct::{Error, Gateway};

fn main() -> ExitCode {
    match run() {
        Ok(()) => ExitCode::SUCCESS,
        Err(e) => {
            eprintln!("error: {e}");
            ExitCode::FAILURE
        }
    }
}

fn run() -> Result<(), Error> {
    let config = std::env::var("LLMUX_CONFIG_JSON").ok();
    let model = std::env::var("LLMUX_MODEL").unwrap_or_else(|_| "demo".to_string());

    let path = llmux::direct::find_library()?;
    println!("library: {}", path.display());

    // `?` here would return without leaking anything: there is no handle yet.
    let gw = Gateway::open_at(&path, config.as_deref())?;
    // From here on the handle is owned by `gw`. It is released by Drop — at the
    // end of this function, on every `?` below, and on a panic. There is no
    // close() to forget, which is the whole reason to wrap a C handle in a Rust
    // type rather than exposing the u64.
    println!("abi:     {}", gw.abi_version());

    // In production, compare that against the version your bindings were
    // generated for and refuse to continue on a mismatch:
    //
    //     let gw = Gateway::open_checked("0.1.2", config.as_deref())?;
    //
    // A shared library resolves off a load path you may not control.

    // ------------------------------------------------------------------ models
    // Answered from memory — no upstream is contacted. This is the call the
    // ffi/ benchmark uses to measure the boundary itself (~4 µs in-process
    // against ~46 µs over loopback).
    let t = Instant::now();
    let models = gw.call("models", None)?;
    println!("models:  {} bytes in {:?}", models.len(), t.elapsed());
    println!("         {}", first_line(&models, 160));

    // -------------------------------------------------------------------- chat
    // The request and the response are the same JSON the HTTP API uses. A body
    // that works against POST /v1/chat/completions works here unchanged.
    let req = format!(
        r#"{{"model":{model},"messages":[{{"role":"user","content":"count to four"}}]}}"#,
        model = json_string(&model)
    );
    let t = Instant::now();
    let chat = gw.call("chat", Some(&req))?;
    println!("chat:    {:?}", t.elapsed());
    println!("         {}", first_line(&chat, 240));

    // A "chat" with "stream": true is REFUSED here rather than quietly served
    // as one blob after the fact. Demonstrating that beats describing it.
    let streaming_req = format!(
        r#"{{"model":{model},"messages":[{{"role":"user","content":"hi"}}],"stream":true}}"#,
        model = json_string(&model)
    );
    match gw.call("chat", Some(&streaming_req)) {
        Ok(_) => println!("refusal: NOT refused — that is a bug in llmux"),
        Err(e) => println!("refusal: {e}"),
    }

    // ------------------------------------------------------------------ stream
    // The idiomatic Rust shape: an Iterator. Under it is a worker thread and a
    // rendezvous channel, so nothing is buffered and time-to-first-token is
    // real. See Gateway::stream.
    print!("stream:  ");
    let t = Instant::now();
    let mut first_token: Option<std::time::Duration> = None;
    let mut chunks = 0usize;
    for chunk in gw.stream(&streaming_req)? {
        let chunk = chunk?; // dropping the iterator here would stop the stream
                            // and join the worker — no leak, no orphan thread.
        chunks += 1;
        if first_token.is_none() {
            first_token = Some(t.elapsed());
        }
        print!("{}", delta_content(&chunk));
    }
    println!();
    println!(
        "         {chunks} chunks, first at {:?}, total {:?}",
        first_token.unwrap_or_default(),
        t.elapsed()
    );

    // ------------------------------------------------------- early termination
    // Returning false from the callback stops the stream. It is NOT an error:
    // you returned false, so you already know it happened. Tokens already
    // served are metered either way.
    let mut seen = 0usize;
    gw.stream_with(&streaming_req, |_chunk| {
        seen += 1;
        seen < 2 // stop after the second chunk
    })?;
    println!("early:   stopped after {seen} chunk(s), Ok(()) — stopping is not a failure");

    // ------------------------------------------------------------ error path
    // An unknown method is a clean error string, not a crash, and the message
    // it allocated is freed by Error::from_c before this line runs.
    match gw.call("no-such-method", Some("{}")) {
        Ok(_) => println!("bogus:   unexpectedly succeeded"),
        Err(e) => println!("bogus:   {e}"),
    }

    // ------------------------------------------------------------ cancel
    // llmux 0.1.5 added `llmux_cancel`, the only way to abandon a blocked call
    // without destroying the whole gateway — that is what llmux_close does,
    // and llmux_close must never be called from inside a callback (it waits
    // for the very call running it). `Gateway::stream`'s iterator already
    // reaches it for you: dropping the iterator below, via `break`, is enough.
    //
    // This needs a fake that can prove something the `fakeupstream` this
    // example otherwise points at cannot: how many chunks the upstream
    // actually produced. `sdks/fake-upstream.py` counts a chunk the moment it
    // is written to a socket and serves the count at GET /generated, so this
    // section spawns its own copy of it rather than reusing `path`'s config.
    // "demo" here, not the outer `model`: this section stands up its own
    // fake-upstream.py with its own config, which always routes exactly
    // "demo" — not whatever LLMUX_MODEL happened to be set to for `gw` above.
    match run_cancellation_demo(&path, "demo") {
        Ok(()) => {}
        Err(msg) => println!("cancel:  skipped ({msg})"),
    }

    Ok(())
    // gw drops here: llmux_close(handle).
}

/// Streams from `sdks/fake-upstream.py` at 100ms/chunk, cancels after 3
/// delivered chunks by dropping the iterator, and prints what the consumer
/// saw against what the upstream actually generated.
///
/// Returns `Err` with a human-readable reason if `python3` is not available —
/// the same "skip loudly" convention `find_library` and its callers use for a
/// missing `libllmux` — rather than failing the whole example over a fixture
/// dependency.
fn run_cancellation_demo(lib_path: &std::path::Path, model: &str) -> std::result::Result<(), String> {
    use std::io::BufRead;
    use std::process::{Command, Stdio};

    if Command::new("python3").arg("--version").output().is_err() {
        return Err("python3 not on PATH; sdks/fake-upstream.py needs it".into());
    }

    let script = std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("fake-upstream.py");
    let mut child = Command::new("python3")
        .arg(&script)
        .args(["--text", "one two three four five six seven eight nine ten"])
        .args(["--chunk-delay-ms", "100"])
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        .spawn()
        .map_err(|e| format!("spawn {}: {e}", script.display()))?;

    // Kills the fake on every return path below, including `?`. An example
    // that leaves a listening python process behind every time someone runs
    // it is exactly the kind of small leak this crate's own Gateway::Drop
    // exists to not have.
    struct KillOnDrop(std::process::Child);
    impl Drop for KillOnDrop {
        fn drop(&mut self) {
            let _ = self.0.kill();
            let _ = self.0.wait();
        }
    }
    let stdout = child.stdout.take().ok_or("fake-upstream.py had no stdout")?;
    let _guard = KillOnDrop(child);
    let mut lines = std::io::BufReader::new(stdout).lines();

    let url = lines
        .next()
        .ok_or("fake-upstream.py exited before printing URL")?
        .map_err(|e| e.to_string())?
        .strip_prefix("URL ")
        .ok_or("expected a URL-prefixed line")?
        .to_string();
    let config = lines
        .next()
        .ok_or("fake-upstream.py exited before printing CONFIG")?
        .map_err(|e| e.to_string())?
        .strip_prefix("CONFIG ")
        .ok_or("expected a CONFIG-prefixed line")?
        .to_string();

    // A second, independent gateway against the counting fake — not the `gw`
    // the rest of this example uses, so this section cannot disturb it.
    let cgw = Gateway::open_at(lib_path, Some(&config)).map_err(|e| e.to_string())?;
    let req = format!(
        r#"{{"model":{model},"messages":[{{"role":"user","content":"count to ten"}}],"stream":true}}"#,
        model = json_string(model)
    );

    let mut seen = 0usize;
    {
        let stream = cgw.stream(&req).map_err(|e| e.to_string())?;
        for chunk in stream {
            chunk.map_err(|e| e.to_string())?; // never "context canceled": we cancelled ourselves
            seen += 1;
            if seen == 3 {
                break; // ChunkStream::drop runs here: llmux_cancel, THEN teardown.
            }
        }
    }
    println!("cancel:  consumer saw {seen} chunks");

    // The handle survives: a fresh call on the SAME handle still works.
    let _ = cgw.call("models", None).map_err(|e| e.to_string())?;
    println!("cancel:  handle survives — a fresh call on it still works");

    let generated = llmux::http::get(
        &format!("{url}/generated"),
        std::time::Duration::from_secs(2),
    )
    .map_err(|e| e.to_string())?;
    println!("cancel:  GET /generated -> {generated}");

    Ok(())
}

/// Minimal JSON string quoting, so the example does not need serde to build a
/// two-field request. Anything real should use serde_json.
fn json_string(s: &str) -> String {
    let mut out = String::with_capacity(s.len() + 2);
    out.push('"');
    for c in s.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
    out
}

/// Pulls `choices[0].delta.content` out of a chunk without a JSON parser. Fine
/// for an example; use serde_json in anything real.
fn delta_content(chunk: &str) -> String {
    let Some(i) = chunk.find(r#""content":""#) else {
        return String::new();
    };
    let rest = &chunk[i + r#""content":""#.len()..];
    let mut out = String::new();
    let mut escaped = false;
    for c in rest.chars() {
        if escaped {
            match c {
                'n' => out.push('\n'),
                't' => out.push('\t'),
                c => out.push(c),
            }
            escaped = false;
        } else if c == '\\' {
            escaped = true;
        } else if c == '"' {
            break;
        } else {
            out.push(c);
        }
    }
    out
}

fn first_line(s: &str, max: usize) -> String {
    let line = s.lines().next().unwrap_or("");
    if line.chars().count() <= max {
        return line.to_string();
    }
    let truncated: String = line.chars().take(max).collect();
    format!("{truncated}…")
}
